From 97e56643d3cac2910f1229db3db6add3f4bef051 Mon Sep 17 00:00:00 2001
From: Dominic Cleal <dominic@cleal.org>
Date: Tue, 5 Apr 2016 10:44:34 +0100
Subject: [PATCH 3/4] refs #14448 - use kafo_parsers .find_available to load
 parser

By using the parser cache, the parser is now optional and will raise an
error at runtime if the cache misses and no parser is available.
---
 lib/kafo/configuration.rb |  2 +-
 lib/kafo/exceptions.rb    |  3 +++
 lib/kafo/puppet_module.rb | 20 +++++++++++++++-----
 3 files changed, 19 insertions(+), 6 deletions(-)

diff --git a/lib/kafo/configuration.rb b/lib/kafo/configuration.rb
index 1dcafa7..7ec20c1 100644
--- a/lib/kafo/configuration.rb
+++ b/lib/kafo/configuration.rb
@@ -86,7 +86,7 @@ module Kafo
     end

     def modules
-      @modules ||= @data.keys.map { |mod| PuppetModule.new(mod, KafoParsers::PuppetModuleParser, self).parse }.sort
+      @modules ||= @data.keys.map { |mod| PuppetModule.new(mod, PuppetModule.find_parser, self).parse }.sort
     end

     def root_dir
diff --git a/lib/kafo/exceptions.rb b/lib/kafo/exceptions.rb
index 1fd7bb9..c1a6e2a 100644
--- a/lib/kafo/exceptions.rb
+++ b/lib/kafo/exceptions.rb
@@ -6,6 +6,9 @@ module Kafo
   class ConditionError < StandardError
   end

+  class ParserError < StandardError
+  end
+
   class TypeError < StandardError
   end
 end
diff --git a/lib/kafo/puppet_module.rb b/lib/kafo/puppet_module.rb
index 84c5429..0c80f11 100644
--- a/lib/kafo/puppet_module.rb
+++ b/lib/kafo/puppet_module.rb
@@ -2,7 +2,7 @@
 require 'kafo/param'
 require 'kafo/param_builder'
 require 'kafo/parser_cache_reader'
-require 'kafo_parsers/puppet_module_parser'
+require 'kafo_parsers/parsers'
 require 'kafo/validator'

 module Kafo
@@ -12,7 +12,11 @@ module Kafo
     attr_reader :name, :identifier, :params, :dir_name, :class_name, :manifest_name, :manifest_path,
                 :groups, :params_path, :params_class_name, :configuration, :raw_data

-    def initialize(identifier, parser = KafoParsers::PuppetModuleParser, configuration = KafoConfigure.config)
+    def self.find_parser
+      KafoParsers::Parsers.find_available(:logger => KafoConfigure.logger)
+    end
+
+    def initialize(identifier, parser = self.class.find_parser, configuration = KafoConfigure.config)
       @identifier        = identifier
       @configuration     = configuration
       @name              = get_name
@@ -49,9 +53,15 @@ module Kafo
     end

     def parse(builder_klass = ParamBuilder)
-      @params      = []
-      @raw_data    = @parser_cache.get(identifier, manifest_path) if @parser_cache
-      @raw_data  ||= @parser.parse(manifest_path)
+      @raw_data = @parser_cache.get(identifier, manifest_path) if @parser_cache
+      if @raw_data.nil?
+        if @parser.nil?
+          raise ParserError.new("No Puppet module parser is installed and no cache of the file #{manifest_path} is available. Please check debug logs and install optional dependencies for the parser.")
+        else
+          @raw_data = @parser.parse(manifest_path)
+        end
+      end
+
       builder      = builder_klass.new(self, @raw_data)
       @validations = @raw_data[:validations]

--
2.4.3
